src/flash.rs

```diff
@@ -2,6 +2,294 @@
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Set lock bit
+        self.flash.cr.cr().modify(|_, w| w.lock().set_bit());
+
+        // Verify success
+        match self.flash.cr.cr().read().lock().bit_is_set() {
+            true => Ok(()),
+            false => Err(Error::LockError),
+        }
+    }
+
+    fn valid_address(&self, offset: u32) -> Result<()> {
+        if FLASH_START + offset > FLASH_END {
+            Err(Error::AddressLargerThanFlash)
+        } else if offset & 0x1 != 0 {
+            Err(Error::AddressMisaligned)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn valid_length(&self, offset: u32, length: usize) -> Result<()> {
+        if offset + length as u32 > self.flash_sz.kbytes() as u32 {
+            Err(Error::LengthTooLong)
+        } else if length & 0x1 != 0 {
+            Err(Error::LengthNotMultiple2)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Erase sector which contains `start_offset`
+    pub fn page_erase(&mut self, start_offset: u32) -> Result<()> {
+        self.valid_address(start_offset)?;
+
+        // Unlock Flash
+        self.unlock()?;
+
+        // Set Page Erase
+        self.flash.cr.cr().modify(|_, w| w.per().set_bit());
+
+        // Write address bits
+        // NOTE(unsafe) This sets the page address in the Address Register.
+        // The side-effect of this write is that the page will be erased when we
+        // set the STRT bit in the CR below. The address is validated by the
+        // call to self.valid_address() above.
+        unsafe {
+            self.flash
+                .ar
+                .ar()
+                .write(|w| w.far().bits(FLASH_START + start_offset));
+        }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit());
+
+        // Wait for operation to finish
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit());
+
+        // Re-lock flash
+        self.lock()?;
+
+        if sr.wrprterr().bit_is_set() {
+            self.flash.sr.sr().modify(|_, w| w.wrprterr().clear_bit());
+            Err(Error::EraseError)
+        } else {
+            if self.verify {
+                // By subtracting 1 from the sector size and masking with
+                // start_offset, we make 'start' point to the beginning of the
+                // page, and similarly 'end' at the end of the page. We do this
+                // because the entire page should have been erased, regardless
+                // of where in the page the given 'start_offset' was.
+                let start = start_offset & !(self.sector_sz.kbytes() as u32 - 1);
+                let end = start + self.sector_sz.kbytes() as u32 - 1;
+                for idx in start..end {
```

Comment:
```suggestion
                // By subtracting 1 from the sector size and masking with
                // start_offset, we make 'start' point to the beginning of the
                // page. We do this
                // because the entire page should have been erased, regardless
                // of where in the page the given 'start_offset' was.
                let size = self.sector_sz.kbytes() as u32;
                let start = start_offset & !(size - 1);
                for idx in start..start+size {
```

Reply:
Is this what you want to do?

Reply:
I like that, I think it is more readable than the previous version.